libftl/ftl-sdk.c

```diff
@@ -417,25 +417,40 @@ FTL_API char* ftl_status_code_to_string(ftl_status_t status) {
 
 BOOL _get_chan_id_and_key(const char *stream_key, uint32_t *chan_id, char *key) {
   size_t len, i = 0;
-  
+
+  char * copy_of_key = _strdup(stream_key);
+
   len = strlen(stream_key);
+
+  // Restream.io stream keys start with 're_' and 3 first letters should be skipped
+  char *restream_stream_key_marker = "re_";
+  if (strncmp(stream_key, restream_stream_key_marker, strlen(restream_stream_key_marker)) == 0) {
+    for (i = 0; i < len - 3; i++) {
+      copy_of_key[i] = copy_of_key[i + 3];
+    }
+
+    copy_of_key[i] = '\0';
+
+    len = strlen(copy_of_key);
+  }
+  free(restream_stream_key_marker);
```

Comment:
@QuinnDamerell-MS Are you sure? Why do we need to free copy_of_key and we don't have to free restream_stream_key_marker? They both are char pointers.

Reply:
Yeah. strdup is calling malloc to allocate a new block of memory in the heap for the string. (https://en.cppreference.com/w/c/experimental/dynamic/strdup). Anytime you allocate from the heap you need to tell the OS you're done with the memory by calling free.

Both `copy_of_key` and `restream_stream_key_marker` are pointers allocated on the stack. Anything on the stack can't be a dynamic size, thus the compiler and allocate and free them when it knows it's safe to. (when the function is being called and when the function is being returned from.) However, `restream_stream_key_marker` is pointing to a read-only block of memory where the static string "re_" is stored. Only one instance of "re_" ever exists in memory per process. It's not allocated by this function, there's just a pointer pointing to it. So you don't need to call free on it.

I'm not even sure what calling free on it will do, I think it will crash.

Reply:
Actually, here's a good post on it. 
https://stackoverflow.com/questions/9504588/should-i-free-char-initialized-using-string-literals